### PR TITLE
Change format of read data

### PR DESCRIPTION
**Pull request by BenjaminRueth**
This PR uses a `dict` instead of a `list` as format for `read_data`. Here the node position is used as keys and the former values of `read_data` are used as values. The new format allows the user to provide additional data or overwrite existing data. The implementation is backwards compatible and allows the user to improve the interpolation with known data, if required.

Another improvement: This feature makes extrapolation for `ExactInterpolationExpression` unnecessary, since we can just provide known data instead of relying on an extrapolation at the "gap" between the coupling interface and the physical boundary (see, for example https://github.com/precice/fenics-adapter/issues/13)

# Todo

- [x] update `PointSource` accordingly
- [ ] write test for `PointSource` @IshaanDesai 
- [x] merge https://github.com/precice/fenics-adapter/pull/81 (this PR is based on #81)
- [x] remove `prototyping` folder
- [x] merge #83 before this one

# Additional information

Originally I implemented this feature to increase the accuracy of the rbf interpolation (the accuracy is actually increased, but not sufficiently). I hoped to get rid of `ExactInterpolationExpression`. However, I realized that the problem with `GeneralInterpolationExpression` lies not only at the borders of the interpolant, but also between the nodes:

Figure_1

The example mimics the temperature at the coupling interface for [`heat.py`](https://github.com/precice/tutorials/blob/95cc1ed1294c9c9e5af4d6bcc6393be4cd8dcb53/HT/partitioned-heat/fenics-fenics/heat.py#L111-L112). 

The observed behaviour is related to the kind of rbf function that [`scipy.interpolate.Rbf`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.Rbf.html) uses. The `multiquadric` function just cannot represent a polynomial function exactly (while `ExactInterpolationExpression` is able to do this and therefore gives more accurate results). An alternative approach that might help us is provided in @uekerman's [thesis, section 4.3.2](https://mediatum.ub.tum.de/doc/1320661/1320661.pdf). However, this will require us to implement the interpolation on our own, as far as I understand it.

### CODE REVIEW

**Comment by BenjaminRueth** [comment]
Related to #74, since extrapolation is unsafe.

**Comment by uekerman** [comment]
There is an RBF alternative which should be easy to realize: to separate the polynomial from the RBF part. First, the polynomial part is fitted with the data, and afterwards the remaining error is interpolated only with the RBF part. So, the RBF interpolation of SciPy could be reused while still capturing the polynomial part exactly. I assume that there is also a least-squares fitting function in SciPy, basically a QR-decomposition.

A similar approach can also be used in preCICE. Mathematical details can be found in [this paper](ftp://ftp.informatik.uni-stuttgart.de/pub/library/ncstrl.ustuttgart_fi/INPROC-2017-35/INPROC-2017-35.pdf), page 6.

**Comment by BenjaminRueth** [comment]
> 
> 
> There is an RBF alternative which should be easy to realize: to separate the polynomial from the RBF part. First, the polynomial part is fitted with the data, and afterwards the remaining error is interpolated only with the RBF part. So, the RBF interpolation of SciPy could be reused while still capturing the polynomial part exactly. I assume that there is also a least-squares fitting function in SciPy, basically a QR-decomposition.
> 
> A similar approach can also be used in preCICE. Mathematical details can be found in this paper, page 6: ftp://ftp.informatik.uni-stuttgart.de/pub/library/ncstrl.ustuttgart_fi/INPROC-2017-35/INPROC-2017-35.pdf .

File: fenicsadapter/fenicsadapter.py [hunk 1]
```diff
@@ -109,21 +109,24 @@ def update_coupling_expression(self, coupling_expression, data):
...
-            for scalar valued data D = 1 and for vector valued data D = dimension of problem.
+        data : dict_like
+            The coupling data. A dictionary containing nodal data with vertex coordinates as key and associated data as
+            value.
         """
-        coupling_expression.update_boundary_data(data, self._coupling_mesh_vertices[:, 0], self._coupling_mesh_vertices[:, 1])
+        x_coord = np.array([vertex[0] for vertex in data.keys()])
+        y_coord = np.array([vertex[1] for vertex in data.keys()])
+        nodal_data = np.array([data for data in data.values()])
+        coupling_expression.update_boundary_data(nodal_data, x_coord, y_coord)
```
Comment (BenjaminRueth) [inline fenicsadapter/fenicsadapter.py:119]: I think we should also add some checks here (dimension fitting? coupling mesh vertices are still present in data?...)

**Comment by BenjaminRueth** [comment]
First experiments with the segregated approach look promising:

segregated

Here is the code, in case anybody is interested:

```
import matplotlib.pyplot as plt
import numpy as np
from scipy.interpolate import Rbf
from scipy.linalg import lstsq

f = lambda x: 4 + 3*x**2
x = np.linspace(0,1,11)
y = np.random.rand(11)
y = f(x)
# a * x ** 2 + b * x + c ~= y
A = np.array([x**2, x, np.ones_like(x)]).T
w, _, _, _ = lstsq(A, y)
lstsq_interp = lambda x: w[0] * x**2 + w[1] * x + w[2]
res = y - lstsq_interp(x)
rbf_interp = Rbf(x,res)
interp = lambda x: rbf_interp(x) + lstsq_interp(x)
xx = np.linspace(0,1,51)
plt.title("segregated RBF interpolation of 4 + 3*x**2")
plt.plot(xx,f(xx),':',label="exact")
plt.plot(xx,interp(xx),label="interp")
plt.plot(x,y,'o',label="nodes")
plt.legend()
plt.show()
```

**Comment by BenjaminRueth** [comment]
https://github.com/precice/fenics-adapter/pull/82/commits/433390bea371837bf2a5614c754bb2ecbb5be137 provides a working version using the segregated RBF interpolation. We can reach a similar level of accuracy as for cubic spline interpolation. Some notes on the current implementation:

* I only provided an implementation for 2d
* The results are slightly changing (coupling iterations etc), since the interpolation schemes (segregated RBF and cubic) work differently. However, the final result is the same within the given accuracy. -> **We will have to update our tests correspondingly.**
* Results for complex domain case are significantly improved (see pictures below)
* offering segregated RBF interpolation has the potential to close multiple issues:
    * #74, since we will not need `ExactInterpolationExpression` anymore
    * #13, since we generalize to the complex geometry case
    * #12, since I do not think we will need linear interpolation anymore

Note: I would suggest to still offer the "old" RBF interpolation, since the segregated version needs a minimum amount of data to create the fit and it is currently only provided for 2d cases.

# Complex case

## "old" RBF interpolation

complex_rbf

## segregated RBF interpolation

complex_rbf_seg

**Comment by BenjaminRueth** [comment]
I decided to split up #82 into two PRs: This one for updating the `read_data` format (#82) and another one, which takes care of adding the new `SegregatedRBFInterpolationExpression` (link will be provided in a moment),

**Comment by IshaanDesai** [comment]
Test for adapter function `get_point_sources` is temporarily disabled [here](https://github.com/precice/fenics-adapter/pull/82/commits/fc2c82d993644254cfd8900d81a670b9aa4637cf#diff-05e5a76b32a33c6ad22c8b60b4c7436fR190). 
The main concern is what do we test while checking the functionality for `get_point_sources`? More details in comment in code.

**Comment by BenjaminRodenberg** [comment]
The results reported in https://github.com/precice/fenics-adapter/pull/82#issuecomment-672896370 cannot be reproduced with the current implementation of the fenics-adapter (https://github.com/precice/fenics-adapter/commit/300ab1a60770c72df0978556bd71e57476b67d69) and the tutorials (https://github.com/precice/tutorials/commit/107c076f2ed65a70cb0cd932387be0c19c0dc667). See #113 for details.

**Pull request merged**